### ISSUE TITLE
None backend - ssh key problems

### ISSUE THREAD

**Issue by spinus** [NixOS/nixops#475]
I can't deploy the box. Nix expression is almost empty. NixOps versions I tried are 1.3.1 and 1.4. Nixpkgs on host machine is unstable, on the virtual machine nixos-version = 16.03.1083.d745044 (Emu).

I can ssh to the box using ssh key.

```
nixops deploy
jenkins-slave1> generating new SSH keypair... done
Enter passphrase for key '/home/spinus/.ssh/id_rsa': 
Host key fingerprint is SHA256:aS+mlf1IQUQRygUL7il89qISapZc1IP4yaDbgFh+JwE
+--[ED25519 256]--+
|      . .o*o     |
|  E  . o =       |
|  ..o . + .      |
| o.+.+ . o       |
|+o= +.* S .      |
|= o=o+.o + .     |
|.=oo o. * +      |
|o=o  . = o o     |
|o  .. .   . .    |
+----[SHA256]-----+
jenkins-slave1> setting state version to 16.03
jenkins-slave1> waiting for SSH................................................
```

So, looks like in the first step nixops could connect to the machine and put ssh key there but in the second step stuff it's not working :[

I can ssh to the box using ssh and nixops ssh

**Comment by domenkozar** [comment]
I saw something similar in the past. I think this is happening due to `_ssh_public_key_deployed` modified in https://github.com/NixOS/nixops/pull/364.

I long knew keeping such states would just cause trouble on the long run. You'll have to find out what really happened there and report it so we can fix it.

**Comment by spinus** [comment]
@domenkozar any idea what's the best solution to debug it? I tried to add --debug but it's not that useful, no debug logging. Best way is to add logging statements to the code or run the code through the debugger or something else?

**Comment by domenkozar** [comment]
@spinus I think it's best to place`import pdb;pdb.set_trace()` inside that code and see what's going on. Debugging isn't verbose enough.

**Comment by spinus** [comment]
Thanks. I'll try.

**Comment by dezgeg** [comment]
Are you by any chance trying to deploy using a hostname that is just an alias in `~/.ssh/config`? I found out that won't work as nixops tries to explicitly that `targetPort` on `targetHost` is open.

**Comment by spinus** [comment]
@dezgeg yes, I'm doing that. I have to do that as I'm connecting to this vm through ssh-proxy/ssh-hop.
It's interesting as ssh command is working, so that would mean that each "command" is implementing ssh connection differently.

**Comment by dezgeg** [comment]
Yes. This is the problematic part: https://github.com/NixOS/nixops/blob/b6e37b7447f85994140acec4093d1067825ec8af/nixops/backends/__init__.py#L262

**Comment by spinus** [comment]
@dezgeg nice, good finding.
@domenkozar, @rbvermaa do you think this would be ok to replace it with equivalent of ```ssh <host> 'exit 0'```?

**Comment by spinus** [comment]
@domenkozar , @rbvermaa would you mind to take a look?

I switched the check to ssh check rather tcp check. As it uses some internal ssh subsystem logging, the output is slightly different than before, it produces some log lines instead of adding dots (I think this should not be the issue probably).

I added timeout as well, but I may remove it if you wish (it's not essential to that change).
